pyecoregen/templates/module.py.tpl

```diff
@@ -20,12 +20,49 @@ eClassifiers = {}
 getEClassifier = partial(Ecore.getEClassifier, searchspace=eClassifiers)
 
 {%- for c in element.eClassifiers if c is type(ecore.EEnum) -%}
+{% if loop.first and textX %}
```

Comment:
I'm not sure I get the point with the `EEnum`. From the quick experiment I made, I didn't get the usage for the `EEnum` redefinition and the `EEnum` instance promotion to a class level. In my test, I just added `MY_ENUM.__name__ = 'MY_ENUM'` for each `EEnum` instance and it seemed to worked. Well, partially, but not because of your code or because of the modification, but because of a bug in PyEcore (I need to patch it quickly), which does not handle `string` attributes as it should when the feature is typed by an `EEnum`.
Should the enumeration be promoted to a class level?

Reply:
textX does not know about enums. An "enum" in textX looks like:
```
BasicType:
	Boolean = /Boolean(?=\W)/ |
	String = /String(?=\W)/ |
	Integer = /Integer(?=\W)/ ;
```
A textX object with value `String` of this rule looks like this:
`
BasicType: {
Boolean: '',
String: 'String',
Integer: '',
}
`

I wanted to create a model that behaves as a textX and an ecore model. To preserve those properties, I used classes. The added `value` property is just syntactical sugar, can be deleted.

Reply:
Your simple `__name__` solution resulted in an error:
```
  File "..\textx\model.py", line 598, in parse_tree_to_objgraph
    model = process_node(parse_tree)
  File "..\textx\model.py", line 397, in process_node
    process_node(n)
  File "..\textx\model.py", line 492, in process_node
    value = process_node(n)
  File "..\textx\model.py", line 397, in process_node
    process_node(n)
  File "..\textx\model.py", line 492, in process_node
    value = process_node(n)
  File "..\textx\model.py", line 397, in process_node
    process_node(n)
  File "..\textx\model.py", line 472, in process_node
    value = process_node(node[0])
  File "..\textx\model.py", line 372, in process_node
    inst = user_class.__new__(user_class)
  File "..\pyecore\ecore.py", line 137, in __new__
    instance = super().__new__(cls)
TypeError: object.__new__(X): X is not a type object (EEnum)
```

pyecore 0.9.0

Reply:
Ok, got it for the `EEnum`! Regarding the simple solution, strange, I must have changed other stuffs during the quick experiment, sorry about that, I'll check this. Don't worry about the syntactical sugar, it can stay, I wasn't sure to understand what was the point. I think that the code could perhaps be simplified a little bit, and there is a collateral effect that should be addressed ;). I quickly test and I tell you about it.

Reply:
@fgro93 I was doing some experiments, but then, re-reading your snippet, it hit me, I'm not sure that 

```
BasicType:
	Boolean = /Boolean(?=\W)/ |
	String = /String(?=\W)/ |
	Integer = /Integer(?=\W)/ ;
```

should be considered an enumeration. This instead should be:

```
BasicType:
	"Boolean" | "String" | "Integer";
```

You case looks more like an `EDataType` mapped to an 'object' instance?

Reply:
True. This is just needed if I want to use a custom class for this rule.
```
BasicType:
	"Boolean" | "String" | "Integer";
```

Problem here: textX won't use our ecore class anymore because `BasicType` does not have any properties. Without properties, an object instantiation is not necessary, therefore custom classes can be ignored.

Result:
```
  File "..\pyecore\valuecontainer.py", line 83, in _set
    self.check(value)
  File "..\pyecore\valuecontainer.py", line 47, in check
    raise BadValueError(value, self.feature.eType)
pyecore.valuecontainer.BadValueError: ("Expected type <class 'mymodel.mymodel.BasicType'>, but got type str with value Boolean instead", 'for class TypeReference')
```

`pyecore.valuecontainer.BadValueError: ("Expected type <class 'mymodel.mymodel.BasicType'>, but got type str with value Boolean instead", 'for class TypeReference')`

Reply:
@aranega please see:
https://github.com/fgro93/pyecore-Provider/tree/enum
 (Attention use the `enum` branch!)

Reply:
@fgro93 Awesome! Thanks for the repository and the examples. Ok, I think the exception you have here is related to the bug in PyEcore. I'm checking this! Also, sorry if I seem picky, I really like your solution of 're-shaping' the PyEcore objects so they match textX and PyEcore, so I really would like to see it dealing with all corner cases ;). Thanks again @fgro93 !

Reply:
I managed to make it work! I mean... The PyEcore bug is still here, let's be honest, but at the end the model is created as it should. To do this, I modified this part of the generated code:

```python
class EEnum(Ecore.EEnum):

    def __init__(self, name=None, default_value=None, literals=None, **kwargs):
        super().__init__(name, default_value, literals)
        self.__dict__.update(kwargs)


BasicTypeENUM = EEnum('BasicType', literals=['Boolean', 'String', 'Integer'])


class BasicType(EEnum):

    def __init__(self, name=None, default_value=None, literals=None, **kwargs):
        super().__init__(**kwargs)
        self.__dict__.update(BasicTypeENUM.from_string(self.value).__dict__)
```

to this:

```
BasicType = EEnum('BasicType', literals=['Boolean', 'String', 'Integer'])
BasicType.__name__ = 'BasicType'
```

And it seemed to be enough. There is perhaps some corner cases however where this fails, it is clear that tests would be required.

Reply:
@aranega Did you change something in pyecore to make this work? I cannot reproduce this as I still get the error: `TypeError: object.__new__(X): X is not a type object (EEnum)`.

I honestly value ecore behavior more than textX behavior, but this approach results in textX `incompatible` obejcts. Not sure if keeping the textX behvior is mandatory here, but I'd love to keep at least the `parent`-child relation.

Reply:
Yes, that was my issue. I used the wrong grammar. My bad, sorry. It works now.

Reply:
I'm relieved :), I was started to think that my machine was the "golden one" where nothing could fail. This make me think that perhaps, a kind of "validator" that could take the grammar, take the metamodel and points the incorrect matches between a rule and a class could be interesting.